crates/bevy_ecs/src/system/system_param.rs

```diff
@@ -216,6 +216,84 @@ impl<'a, T: Resource> FetchSystemParam<'a> for FetchResMut<T> {
     }
 }
 
+pub struct FetchOptionRes<T>(PhantomData<T>);
+
+impl<'a, T: Resource> SystemParam for Option<Res<'a, T>> {
+    type Fetch = FetchOptionRes<T>;
+}
+
+impl<'a, T: Resource> FetchSystemParam<'a> for FetchOptionRes<T> {
+    type Item = Option<Res<'a, T>>;
+
+    fn init(system_state: &mut SystemState, _world: &World, _resources: &mut Resources) {
+        if system_state.resource_access.is_write(&TypeId::of::<T>()) {
+            panic!(
+                "System `{}` has a `Res<{res}>` parameter that conflicts with \
+                another parameter with mutable access to the same `{res}` resource.",
+                system_state.name,
+                res = std::any::type_name::<T>()
+            );
+        }
+        system_state.resource_access.add_read(TypeId::of::<T>());
+    }
+
+    #[inline]
+    unsafe fn get_param(
+        _system_state: &'a SystemState,
+        _world: &'a World,
+        resources: &'a Resources,
+    ) -> Option<Self::Item> {
+        if resources.contains::<T>() {
```

Comment:
This double lookup is not good - ideally we'd just have a `get_unsafe_ref` which returns an `Option<NonNull<T>>`.

(I think the most likely response from cart will be to point this out)